### PR TITLE
New dump format for Ultralight/EV1/NTAG support

### PR DESCRIPTION
**Pull request by McEloff**
1. Mifare Classic 7 byte uid identification from dump
2. Ultralight EV1/NTAG new dump format supports (used in proxmark3)

### CODE REVIEW

File: ChameleonMiniGUI/Json/MifareUltralightCardInfo.cs [hunk 1]
```diff
@@ -30,18 +31,30 @@ public override string Uid
         public string TBO_0 { get; set; }
 
         [DataMember(Order = 3)]
-        public string Tearing { get; set; }
+        public string Pack { get; set; }
```
Comment (iceman1001) [inline ChameleonMiniGUI/Json/MifareUltralightCardInfo.cs]: Where do you see Pack in the MFU json format created in Proxmark3 client?
ref
https://github.com/RfidResearchGroup/proxmark3/blob/master/client/loclass/fileutils.c#L226

File: ChameleonMiniGUI/FrmMain.cs [hunk 2]
```diff
@@ -2141,12 +2152,12 @@ internal void DownloadAndSaveDump(string filename)
 
             // Also check if the tag is UL to save the counters too
             var configStr = SendCommand($"CONFIG{_cmdExtension}?").ToString();
-            if (!string.IsNullOrWhiteSpace(configStr) && (configStr.Contains("ULTRALIGHT")))
+            if (!string.IsNullOrWhiteSpace(configStr) && (configStr.Contains("ULTRALIGHT_EV1")))
```
Comment (iceman1001) [inline ChameleonMiniGUI/FrmMain.cs]: Will this not break backward compability?

File: ChameleonMiniGUI/Json/MifareUltralightModel.cs [hunk 3]
```diff
@@ -37,32 +37,148 @@ public MifareUltralightCardInfo Card
...
+                    .Concat(stba(info.TBO_1))
+                    .Concat(pages)
+                    .Concat(stba(info.Signature))
+                    .Concat(stba(info.Counter0)).Concat(stba(info.Tearing0))
+                    .Concat(stba(info.Counter1)).Concat(stba(info.Tearing1))
+                    .Concat(stba(info.Counter2)).Concat(stba(info.Tearing2))
+                    .Concat(Blocks.SelectMany(bytes => bytes))
+                    .ToArray();
+            }
+            else if (info.Tearing0 != null && info.Tearing1 != null && info.Tearing2 != null)
```
Comment (iceman1001) [inline ChameleonMiniGUI/Json/MifareUltralightModel.cs]: Is this suppose to a detection of format used before?  or`?  Kind of unclear why you do these different choices.

File: ChameleonMiniGUI/Json/MifareUltralightModel.cs [hunk 4]
```diff
@@ -37,32 +37,148 @@ public MifareUltralightCardInfo Card
...
             var stba = new Func<string, byte[]>(MifareClassicModel.StringToByteArray);
-            return stba(info.Version)
-                .Concat(stba(info.TBO_0))
-                .Concat(stba(info.Tearing))
-                .Concat(stba(info.Pack))
-                .Concat(stba(info.TBO_1))
-                .Concat(stba(info.Signature))
-                .Concat(Blocks.SelectMany(bytes => bytes))
-                .ToArray();
+            if (info.Counter0 != null && info.Counter1 != null && info.Counter2 != null) {
```
Comment (iceman1001) [inline ChameleonMiniGUI/Json/MifareUltralightModel.cs]: Is this suppose to a detection of format used before?  or`?  Kind of unclear why you do these different choices.

File: ChameleonMiniGUI/Json/MifareUltralightModel.cs [hunk 5]
```diff
@@ -37,32 +37,148 @@ public MifareUltralightCardInfo Card
...
+            bool isNewHeaderFormat = HasUltralightNewHeader(data);
+            bool isOldHeaderFormat = !isNewHeaderFormat && HasUltralightHeader(data);
+            int headerLength = isNewHeaderFormat ? MifareUltralightCardInfo.NewPrefixLength : isOldHeaderFormat ? MifareUltralightCardInfo.PrefixLength : 0;
+            int pages = (data.Length - headerLength) / 4;
             return new MifareUltralightModel()
             {
                 Created = "ChameleonMiniGUI",
-                Card = new MifareUltralightCardInfo()
+                Card = isNewHeaderFormat ?
+                new MifareUltralightCardInfo()
```
Comment (iceman1001) [inline ChameleonMiniGUI/Json/MifareUltralightModel.cs]: Make this return of object nice and clean.  Not nested like this.

File: ChameleonMiniGUI/Json/MifareUltralightCardInfo.cs (hunk 1 above)
Comment (McEloff) [reply ChameleonMiniGUI/Json/MifareUltralightCardInfo.cs]: removed

File: ChameleonMiniGUI/FrmMain.cs (hunk 2 above)
Comment (McEloff) [reply ChameleonMiniGUI/FrmMain.cs]: It is used to download from chameleon some extra pages (52 bytes) with extra data. Simple ultralights has no extra data, no need to download more bytes

File: ChameleonMiniGUI/Json/MifareUltralightModel.cs (hunk 3 above)
Comment (McEloff) [reply ChameleonMiniGUI/Json/MifareUltralightModel.cs]: removed

File: ChameleonMiniGUI/Json/MifareUltralightModel.cs (hunk 4 above)
Comment (McEloff) [reply ChameleonMiniGUI/Json/MifareUltralightModel.cs]: removed

File: ChameleonMiniGUI/Json/MifareUltralightModel.cs (hunk 5 above)
Comment (McEloff) [reply ChameleonMiniGUI/Json/MifareUltralightModel.cs]: fixed

**Comment by iceman1001** [comment]
Thanks @McEloff !

**Pull request merged**